File: leetcode/380-insert-delete-getrandom-o1/main.go
package main

import (
	"math/rand"
	"time"
)

/*
	questions to ask:
	- each item means each the items inserted or the key inserted? allow duplicate?

	1st approach: hashtable

	Insert Time				O(1)
	Remove Time				O(1)
	GetRandom Time		O(n)
	Space							O(n) the unique keys
	1336 ms, faster than 6.52%
*/

type RandomizedSet struct {
	HashTable map[int]bool
}

/** Initialize your data structure here. */
func Constructor() RandomizedSet {
	return RandomizedSet{make(map[int]bool)}
}

/** Inserts a value to the set. Returns true if the set did not already contain the specified element. */
func (this *RandomizedSet) Insert(val int) bool {
	if _, x := this.HashTable[val]; x {
		return false
	}
	this.HashTable[val] = true
	return true
}

/** Removes a value from the set. Returns true if the set contained the specified element. */
func (this *RandomizedSet) Remove(val int) bool {
	if _, x := this.HashTable[val]; x {
		delete(this.HashTable, val)
		return true
	}
	return false
}

/** Get a random element from the set. */
func (this *RandomizedSet) GetRandom() int {
	rand.Seed(time.Now().UTC().UnixNano())
	n := len(this.HashTable)
	targetIdx := rand.Intn(n)
	keys := []int{}
	for k, _ := range this.HashTable {
		keys = append(keys, k)
	}
	key := keys[targetIdx]
	return key
}

func main() {
}

File: leetcode/380-insert-delete-getrandom-o1/main.py
import random

"""
    1st: hashtable
    - straight forward approach

    Insert Time				O(1)
	Remove Time				O(1)
	GetRandom Time		    O(N)
	Space					O(n) the unique keys
    256 ms, faster than 29.35% 
"""


class RandomizedSet(object):

    def __init__(self):
        self.ht = {}

    def insert(self, val):
        if val in self.ht:
            return False
        self.ht[val] = True
        return True

    def remove(self, val):
        if val not in self.ht:
            return False
        del self.ht[val]
        return True

    def getRandom(self):
        idx = random.randint(0, len(self.ht)-1)
        nums = self.ht.keys()
        return nums[idx]


"""
	2nd approach: hashtable + array, learned from others
	- save value: index in hashtable
	- when delete, swap the target item and the last item in the array, and remove the last item
	- see ./idea_add.png and ./idea_remove.png

	Insert Time				O(1)
	Remove Time				O(1)
	GetRandom Time		    O(1)
	Space					O(n) the unique keys
	92 ms, faster than 78.93%
"""


class RandomizedSet(object):

    def __init__(self):
        self.ht = {}
        self.nums = []

    def insert(self, val):
        if val in self.ht:
            return False
        self.ht[val] = len(self.nums)
        self.nums.append(val)
        return True

    def remove(self, val):
        if val not in self.ht:
            return False
        idx = self.ht[val]
        last = self.nums[-1]
        self.ht[last] = idx
        self.nums[idx] = last
        self.nums.pop()
        del self.ht[val]
        return True

    def getRandom(self):
        idx = random.randint(0, len(self.nums)-1)
        return self.nums[idx]
